Gregsdennis/locatable iri adjustments (#1435)
json-schema-org/json-schema-spec: jsonschema-core.md

Note that this IRI is an identifier and not necessarily a network locator. In
the case of a network-addressable URL, a schema need not be downloadable from
its canonical IRI. However, schema authors SHOULD refrain from creating schemas
that use a locatable `$id` value unless the schema will be available at the
indicated location. Otherwise, an IRI which does not convey location semantics

> SHOULD be used.

Author: I'm still iffy on including this.  We don't have much in the way for recommendations for schema authors mainly because we really can't control what they do.

Reviewer: I'm not convinced we should include this in the spec either.
I think this is the only section which strays into best practice though.

> I also don't think this recommendation makes sense given the direction these changes and the discussions behind them have gone.

I think we have (speaking personally) intrenched to almost always use https, that using a non-locating semantics URI scheme seems strange. But I do think it's something we should strive to wards suggesting as best practice. It would avoid confusion, as you've said. But I also think we shouldn't allow things to be required that way. (Won't say any more on the topic here.)

I think we can/should track potential best practice recomendations as Issues in the Community repo, and maybe migrating them to the new website repo when appropriate.

The resolved IRI produced by these keywords is not necessarily a network
locator, only an identifier. A schema need not be downloadable from the address
if it is a network-addressable URL. Implementations which can access the network

> MUST default to operating offline.

Reviewer 2: This needs to be a "SHOULD" to not contradict the "SHOULD" on 1285.

downloaded, but instead JSON Schema implementations SHOULD understand ahead of
time which schemas they will be using, and the IRIs that identify them.

Implementations SHOULD be able to associate arbitrary IRIs with an arbitrary
schema and/or automatically associate a schema's `$id`-given IRI, depending on
the trust that the validator has in the schema. Such IRIs and schemas can be
supplied to an implementation prior to processing instances, or may be noted
within a schema document as it is processed, producing associations as shown in
{{idexamples}}.

> Implementations MAY provide functionality to automatically fetch schemas based
> on location semantics expressed by the URI, however such functionality MUST be
> disabled by default to prefer offline operation. When schemas are downloaded,
> for example by a generic user-agent that does not know until runtime which
> schemas to download, see {{hypermedia}}.

Reviewer 2: > such functionality MUST be disabled by default to prefer offline operation

This would mean you can't write a hyper-schema client that functions at all in it's default configuration. We can't do that. It also contradicts line 1137 that says this is a "SHOULD". Saying it once on that line should be enough. I also don't think strengthening this "SHOULD" to a "MUST" accurately reflects the opinions expressed in the discussion.

Here's an alternative that I think avoids the useless-hyper-schema-client problem and reflects the opinions expressed in the discussion.

```suggestion
Some types of implementations, such as generic user-agents (see {{hypermedia}}), don't know what schemas they will need until runtime. Those implementations MAY automatically fetch schemas based on location semantics expressed by the URI.
```

We aren't explicitly giving permission to fetch schemas to any and all types of implementations, just ones where that's the only mode in which the implementation could possibly work. This doesn't apply to typical validator implementations that can know schemas ahead of time. The recommended behavior of those types of implementations is given on line 1137.

Reviewer: I think the proposed suggestion above is fine.

Author: He's right that it conflicts with 1137, though.  I'm happy to make the MUST a SHOULD, but I think I prefer the wording that's there.

Reviewer 3: How does it conflict?  This line just says that the config option can exist; it doesn't say that it's allowed to default to on.

I would prefer to say that the config option to download, if it exists at all, MUST default to off. Otherwise we have security considerations that must be documented more visibly.

Author: From "Schema References" section

> Implementations which can access the network SHOULD default to operating offline.

Then this text said

> ... such functionality MUST be disabled by default to prefer offline operation.

Author: Regarding the redudancy, I'm happy leaving it as it's pertinent to both cases, and I don't feel that it's necessary to reference an entire section when all we need from it is a single sentence.  I think the duplication is fine in this case.

Author: > Otherwise we have security considerations that must be documented more visibly.

We are also looking at (separately) augmenting the security considerations section to address this.

Reviewer 3: Yes, duplication in a single line is ok as long as it's consistent.
I'm still strongly in favour of "MUST default to offline" though.

Author: I'm inclined to agree that MUST is best here.

> This would mean you can't write a hyper-schema client that functions at all in it's default configuration. - @jdesrosiers 

I wouldn't think a hyper-schema client is strictly a JSON Schema implementation.  I would say that it _uses_ a JSON Schema implementation, even if that implementation is built in.  Since that's the case, the client would need to configure the schema implementation to access the network.

These requirements really only apply to JSON Schema implementations that are exposed as such.  If an implementation is built into some other functionality (e.g. a hyper-schema client) specifically to serve that functionality, then it makes sense for the implementation to take exception to things that hinder that functionality.

Thus, if your hyper-schema client needs its JSON Schema implementation to access the internet by default, go for it.  It's not violating the spec for doing so because no one is using it _as_ a JSON Schema implementation.

On the other hand, if you expose it _as_ a JSON Schema implementation, then when it's operating in that capacity, it needs to follow these requirements.

Reviewer 2: I see what you're saying, but I think your definition of what constitutes an "implementation" is more narrow than what our charter indicates. Validators are not the only type of implementation this spec needs to support. We intend to enable and support other types of implementations that are not validators such as code generators, form generators, documentation generators, hyper-schema clients, and more that haven't been thought of yet.

The way I see it, the "core" spec is a common set of rules that applies to all types of JSON Schema implementations including non-validator implementations like hyper-schema. The "validation" spec is specific to validator implementations. There's currently a little cross over in that the applicator vocabulary is defined in "core", but that was done specifically because it was thought that any type of implementation would need applicators.

So, I can't agree that a hyper-schema client is not a JSON Schema implementation. Although it uses a validation implementation it is also in itself independently a non-validator JSON Schema implementation.

Reviewer 2: > Regarding the redudancy, I'm happy leaving it as it's pertinent to both cases, and I don't feel that it's necessary to reference an entire section when all we need from it is a single sentence. I think the duplication is fine in this case.

I don't see that these are two distinct cases. I think we're just saying the same thing in two different places. It doesn't need to be said twice.

Reviewer 2: > I think I prefer the wording that's there.

If you don't like the way I phrased it, that's fine, but it really needs to somehow acknowledge that the requirement is nonsense in some applications and that the requirement doesn't apply in those cases.

Author: > I think your definition of what constitutes an "implementation" is more narrow than what our charter indicates.

https://github.com/json-schema-org/community/pull/325/files#diff-91d16dd057bd39f6431e90a1d78e9224eafafcedc611eba0878d77e6fda86601R10

> JSON Schema aims to enable the confident and reliable use of the JSON data format. It does this primarily by providing specification documents which define a declarative language that allows annotation and validation of JSON documents.

The (proposed) charter clearly states that the primary purposes are "annotation and validation."  Hypermedia and other concerns are explicitly listed as secondary.

> We intend to enable and support other types of implementations that are not validators such as code generators, form generators, documentation generators, hyper-schema clients, and more that haven't been thought of yet.

But not at the cost of compromising the integrity of the primary concerns.  Being able to reach out to the internets or access the file system is not a requirement of annotation or validation.  The current text _recognizes_ the need by permitting implementations to have an online mode, but I maintain that the default behavior must be offline.

Hyperschema, codegen, and all of those other uses are _uses_ of JSON Schema.  They are not targets.  I will not weaken the spec to accommodate secondary and tertiary uses for the spec.  Strictly speaking, none of these are JSON Schema implementations.  They _use_ JSON Schema.  As such, JSON Schema only needs to be defined to serve its primary purposes, and users of JSON Schema (including secondary specifications) need to work within those bounds.

Reviewer 2: We should all agree what constitutes an "implementation". Let's discuss that topic at the next OCWM.

Author: I think for the purposes of this, we're on the same page on the definition of an implementation.  I think MUST is more appropriate, but I'm okay with this being a SHOULD.  Technically the hyper-schema bits can annotate an instance offline, but to do so outside of the context of a client (which definitely needs online access) doesn't really seem useful.

Reviewer 2: > Technically the hyper-schema bits can annotate an instance offline

Fetching a schema can't be completely decoupled from evaluation of the schema in the case of hyper-schema. The schema may include references to other schemas that also need to be fetched. You don't know about those additional schemas until you are evaluating the schema, so it's necessary for the fetch functionality to be integrated with the evaluation process.

JSON Schema output is defined using the JSON Schema data instance model as
described in section 4.2.1. Implementations MAY deviate from this as supported
by their specific languages and platforms, however it is RECOMMENDED that the

> output be convertib le to the JSON format defined herein via serialization or

Reviewer 3: ```suggestion
output be convertible to the JSON format defined herein via serialization or
```

Author: Nice catch, yeah.  The fix for that came in when I rebased just now.
